### PR TITLE
Add a SlackExceptionHandler.

### PR DESCRIPTION
**Pull request by garnertb**
I wanted to see if there is room for this functionality in your project.  It re-configures the Django mail_admins log handler from Django to send logs to a slack channel:

screen shot 2015-10-14 at 11 15 05 am

Simply add the handler to the LOGGING dict and you can get nice slack messages.

```
'slack_admins': {
            'level': 'ERROR',
            'filters': ['require_debug_false'],
            'class': 'django_slack.logs.SlackExceptionHandler'
        }
```

and 
```
"django": {
            "handlers": ["console", "slack_admins"], "level": "ERROR", },
```

### CODE REVIEW

File: django_slack/logs.py [hunk 1]
```diff
@@ -0,0 +1,80 @@
...
+
+        if record.exc_info:
+            exc_info = record.exc_info
+        else:
+            exc_info = (None, record.getMessage(), None)
+
+        reporter = ExceptionReporter(request, is_email=True, *exc_info)
+        message = "%s\n\n%s" % (self.format(no_exc_record), reporter.get_traceback_text())
+
+        colors = {'ERROR': 'danger',
```
Comment (lamby) [inline django_slack/logs.py]: Please flow these like so:

```
a = {
    'b': 'c',
    'd': 'e',
}
```

.. to match the rest of the project :)

File: django_slack/logs.py [hunk 2]
```diff
@@ -0,0 +1,80 @@
...
+                  'WARNING': 'warning',
+                  'INFO': 'good'}
+
+        attachments = {'title': subject,
+                       'text': message,
+                       'color': colors.get(record.levelname, '#AAAAAA')}
+
+        attachments.update(self.kwargs)
+        self.send_message(self.template, {'text': subject}, self.generate_attachments(**attachments))
```
Comment (lamby) [inline django_slack/logs.py]: Long-line; please wrap?

File: django_slack/templates/django_slack/exception [hunk 3]
```diff
@@ -0,0 +1,5 @@
+{%extends django_slack%}
```
Comment (lamby) [inline django_slack/templates/django_slack/exception]: Please add spaces between `{%` and the start (and `{{` below too)

File: django_slack/logs.py [hunk 4]
```diff
@@ -0,0 +1,80 @@
...
+from django_slack import slack_message
+
+
+class SlackExceptionHandler(logging.Handler):
+    """
+    An exception log handler that sends log entries to a slack channel.
+    """
+
+    def __init__(self, **kwargs):
+        # pop any kwargs that shouldn't be passed into the slack message attachment here.
```
Comment (lamby) [inline django_slack/logs.py]: Don't quite follow this - could you eleborate in the comment?

File: django_slack/logs.py [hunk 5]
```diff
@@ -0,0 +1,80 @@
+import logging
+from copy import copy
+from django.conf import settings
+from django.views.debug import ExceptionReporter
+from django_slack import slack_message
+
+
+class SlackExceptionHandler(logging.Handler):
+    """
+    An exception log handler that sends log entries to a slack channel.
```
Comment (lamby) [inline django_slack/logs.py]: slack → Slack

File: django_slack/logs.py [hunk 6]
```diff
@@ -0,0 +1,80 @@
+import logging
+from copy import copy
+from django.conf import settings
+from django.views.debug import ExceptionReporter
+from django_slack import slack_message
```
Comment (lamby) [inline django_slack/logs.py]: This can be a relative import (ie. `from . import slack_message`) I think ?

File: django_slack/logs.py [hunk 7]
```diff
@@ -0,0 +1,80 @@
+import logging
+from copy import copy
```
Comment (lamby) [inline django_slack/logs.py]: Please import `copy` then use `copy.copy`

**Comment by lamby** [comment]
Like this. Please could you:

 * address the cosmetic changes I've just spammed you with
 * update the documentation? having the installation instructions in a pull request isn't ideal and will get lost and out of date

:)

**Comment by garnertb** [comment]
@lamby, I've implemented the suggestions.  Let me know if there is anything else.

File: django_slack/logs.py [hunk 8]
```diff
@@ -0,0 +1,84 @@
...
+from django.views.debug import ExceptionReporter
+
+
+class SlackExceptionHandler(logging.Handler):
+    """
+    An exception log handler that sends log entries to a Slack channel.
+    """
+
+    def __init__(self, **kwargs):
+        # pop any kwargs that shouldn't be passed into the slack message attachment here.
```
Comment (lamby) [inline django_slack/logs.py]: Still don't really understand this. I mean, I understand the *words* but in about a week I will have no idea why it (what?) is being passed in. And why it shouldn't be passed to the attachment.

Also, slack → Slack :)

File: django_slack/logs.py [hunk 9]
```diff
@@ -0,0 +1,84 @@
+import copy
+import logging
+from . import slack_message
+from django.conf import settings
+from django.views.debug import ExceptionReporter
+
```
Comment (lamby) [inline django_slack/logs.py]: why the double-space?

File: django_slack/logs.py [hunk 10]
```diff
@@ -0,0 +1,84 @@
...
+        logging.Handler.__init__(self)
+
+    def emit(self, record):
+        try:
+            request = record.request
+            subject = '%s (%s IP): %s' % (
+                record.levelname,
+                ('internal' if request.META.get('REMOTE_ADDR') in settings.INTERNAL_IPS
+                 else 'EXTERNAL'),
+                record.getMessage()
```
Comment (lamby) [inline django_slack/logs.py]: Missing trailing semicolon - adding a new bit of text will add unnecessary diff noise

File: django_slack/logs.py [hunk 11]
```diff
@@ -0,0 +1,84 @@
...
+            )
+        except Exception:
+            subject = '%s: %s' % (
+                record.levelname,
+                record.getMessage()
```
Comment (lamby) [inline django_slack/logs.py]: (ditto)

File: django_slack/logs.py [hunk 12]
```diff
@@ -0,0 +1,84 @@
...
+    def emit(self, record):
+        try:
+            request = record.request
+            subject = '%s (%s IP): %s' % (
+                record.levelname,
+                ('internal' if request.META.get('REMOTE_ADDR') in settings.INTERNAL_IPS
+                 else 'EXTERNAL'),
+                record.getMessage()
+            )
+        except Exception:
```
Comment (lamby) [inline django_slack/logs.py]: Isn't this too broad?

File: django_slack/logs.py [hunk 13]
```diff
@@ -0,0 +1,84 @@
...
+        else:
+            exc_info = (None, record.getMessage(), None)
+
+        reporter = ExceptionReporter(request, is_email=True, *exc_info)
+        message = "%s\n\n%s" % (self.format(no_exc_record), reporter.get_traceback_text())
+
+        colors = {
+            'ERROR': 'danger',
+            'WARNING': 'warning',
+            'INFO': 'good'
```
Comment (lamby) [inline django_slack/logs.py]: semicolon

File: django_slack/logs.py [hunk 14]
```diff
@@ -0,0 +1,84 @@
...
+        }
+
+        attachments = {
+            'title': subject,
+            'text': message,
+            'color': colors.get(record.levelname, '#AAAAAA')}
```
Comment (lamby) [inline django_slack/logs.py]: Why touch the `colors` layout but not this one? :) (last line, etc.)

File: django_slack/templates/django_slack/exception [hunk 15]
```diff
@@ -0,0 +1,5 @@
+{% extends django_slack %}
```
Comment (lamby) [inline django_slack/templates/django_slack/exception]: is this right? Look at the other templates, they are just ``{% block blah %}{% endblock %}``

(This is really another way of saying, please comment this exception!)

File: django_slack/logs.py (hunk 9 above)
Comment (garnertb) [inline django_slack/logs.py]: PEP8, two spaces before each class.

File: django_slack/logs.py (hunk 10 above)
Comment (garnertb) [inline django_slack/logs.py]: I kept this as it was implemented in the Django AdminEmailHandler, but can add a trailing comma if preferred.

File: django_slack/logs.py (hunk 12 above)
Comment (garnertb) [inline django_slack/logs.py]: I kept this as it was implemented in the Django AdminEmailHandler.

File: django_slack/logs.py (hunk 11 above)
Comment (garnertb) [inline django_slack/logs.py]: I kept this as it was implemented in the Django AdminEmailHandler, but can add a trailing comma if preferred.

**Comment by garnertb** [comment]
@lamby, thanks again for the feedback.  I've implemented the changes.

**Comment by lamby** [comment]
Merged, many thanks. Will release in a sec

**Comment by lamby** [comment]
I have made a few changes and am still not 100% happy with how the templating is commented - the comment you added to the file under templates/ didn't impart any information tbh.

It's not even a message "part", so having it in that dir is completely wrong tbh

**Comment by lamby** [comment]
Yeah, 347b73f makes it so much easier to follow - it's not part of the template hack files..

**Comment by lamby** [comment]
Also renamed "logs.py" -> "log.py" btw.

Released!

**Comment by garnertb** [comment]
@lamby, thanks for the clean up work, much appreciated.
